include/cuda/memory_resource

```diff
@@ -0,0 +1,932 @@
+
+  virtual ~memory_resource() = default;
+
+  /*!
+   * \brief Returns the resource's execution context
+   *
+   * Inherited from base class
+   *
+   * context get_context() const noexcept;
+   */
+  context get_context() const noexcept { return this->do_get_context(); }
+
+  /*!
+   * \brief Allocates storage of size at least `__bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `__alignment` if such
+   * alignment is supported. Otherwise throws.
+   *
+   * Storage may be accessed immediately within the execution context returned
+   * by `get_context()`, otherwise synchronization is required.
+   *
+   * \throws If storage of the requested size and alignment cannot be obtained.
+   *
+   * \param __bytes The size in bytes of the allocation
+   * \param __alignment The alignment of the allocation
+   * \return Pointer to the requested storage
+   */
+  void *allocate(std::size_t __bytes,
+                 std::size_t __alignment = default_alignment) {
+    return do_allocate(__bytes, __alignment);
+  }
+
+  /*!
+   * \brief Deallocates the storage pointed to by `__p`.
+   *
+   * `__p` must have been returned by a prior call to `allocate(__bytes,
+   * __alignment)` on a `memory_resource` that compares equal to `*this`, and
+   * the storage it points to must not yet have been deallocated, otherwise
+   * behavior is undefined.
+   *
+   * \throws Nothing.
+   *
+   * \param __p Pointer to storage to be deallocated
+   * \param __bytes The size in bytes of the allocation. This must be equal to
+   * the value of `__bytes` that was specified to the `allocate` call that
+   * returned `__p`.
+   * \param __alignment The alignment of the allocation. This must be equal to
+   * the value of `__alignment` that was specified to the `allocate` call that
+   * returned `__p`.
+   */
+  void deallocate(void *__p, std::size_t __bytes,
+                  std::size_t __alignment = default_alignment) {
+    do_deallocate(__p, __bytes, __alignment);
+  }
+
+  /*!
+   * \brief Compare this resource to another.
+   *
+   * Two resources compare equal if and only if memory allocated from one
+   * resource can be deallocated from the other and vice versa.
+   *
+   * \param __other The other resource to compare against
+   */
+  bool is_equal(memory_resource const& __other) const noexcept {
+    return do_is_equal(__other);
+  }
+
+private:
+  template <typename, typename...>
+  friend class basic_resource_view;
+
+  void *do_allocate(std::size_t __bytes, std::size_t __alignment) override = 0;
+
+  void do_deallocate(void *__p, std::size_t __bytes,
+                     std::size_t __alignment) override = 0;
+
+  // Default to identity comparison
+  virtual bool do_is_equal(memory_resource const &__other) const noexcept {
+    return this == &__other;
+  }
+
+  virtual void *__do_as_kind(const ::std::type_info *__kind_type_id, const void *__kind_type_fallback_id) const noexcept {
```

Comment:
I think this should be final override? Or at least override -- clangd complains.

```suggestion
  virtual void *__do_as_kind(const ::std::type_info *__kind_type_id, const void *__kind_type_fallback_id) const noexcept final override {
```

Reply:
Yes, it is the one of few places where `final` is a good idea.

Reply:
Done. According to c++ core guidelines, `final` and `override` should not be used together, since both of them can only be specified on an overridden virtual function, making `override` redundant in presence of `final`.

Reply:
OK. You had `final override` on another function, so make sure you update that one too. I don't think I've uttered the phrase "final override" in my life before this thread. :)

Reply:
I've fixed that one, too.